galaxy/worker/tasks/collection.py

```diff
@@ -0,0 +1,143 @@
+
+import logging
+import tempfile
+import tarfile
+
+from django.db import transaction
+from django.db.utils import IntegrityError
+from pulpcore.app import models as pulp_models
+
+from galaxy.common import schema
+from galaxy.importer import collection as importer
+from galaxy.importer import exceptions as i_exc
+from galaxy.main import models
+from galaxy.worker import exceptions as exc
+from galaxy.worker import logutils
+
+
+log = logging.getLogger(__name__)
+
+
+def import_collection(artifact_id, repository_id):
+    task = models.CollectionImport.current()
+    log.info('Starting collection import task: {}'.format(task.id))
+
+    artifact = pulp_models.Artifact.objects.get(pk=artifact_id)
+    repository = pulp_models.Repository.objects.get(pk=repository_id)
+
+    filename = schema.CollectionFilename(
+        task.namespace.name, task.name, task.version)
+
+    task_logger = _get_task_logger(task)
+    task_logger.info('Starting import: task_id={}, artifact_id={}'
+                     .format(task.id, artifact_id))
+    try:
+        collection_info = _process_collection(
+            artifact, filename, task_logger)
+        _publish_collection(task, artifact, repository, collection_info)
+    except exc.TaskError as e:
+        artifact.delete()
+        task_logger.error('Import Task "{task_id}" failed: {message}'
+                          .format(task_id=task.id, message=str(e)))
+        raise
+
+    warnings, errors = task.get_message_stats()
+    msg = ('Import completed with {warnings} warnings and {errors} errors'
+           .format(warnings=warnings, errors=errors))
+    task_logger.info(msg)
+
+
+def _get_task_logger(task):
+    logger = logging.getLogger('galaxy.worker.tasks.import_collection')
+    return logutils.ImportTaskAdapter(logger, task=task)
+
+
+def _process_collection(artifact, filename, task_logger):
+    with tempfile.TemporaryDirectory() as extract_dir:
+        with artifact.file.open() as pkg_file, \
+                tarfile.open(fileobj=pkg_file) as pkg_tar:
+            pkg_tar.extractall(extract_dir)
+
+        try:
+            collection_info = importer.import_collection(
+                extract_dir, filename, task_logger)
+        except i_exc.ImporterError as e:
+            raise exc.ImportFailed(str(e))
+
+    _log_collection_info(collection_info)
+    return collection_info
+
+
+@transaction.atomic
+def _publish_collection(task, artifact, repository, collection_info):
+    metadata = collection_info.collection_info
+    collection, _ = models.Collection.objects.get_or_create(
+        namespace=task.namespace, name=metadata.name)
+
+    try:
+        version = collection.versions.create(
+            version=metadata.version,
+            metadata=metadata.get_json(),
+            quality_score=collection_info.quality_score,
+        )
+    except IntegrityError:
+        raise exc.VersionConflict('Collection version already exists.')
+
+    relative_path = '{0}-{1}-{2}.tar.gz'.format(
```

Comment:
Could use something like

``` python
relative_path = '{metadata.namespace}-{metadata.name}-{metadata.version}.tar.gz'.format(metadata=metadata)
```

Not sure if it matters much for this case, so either is fine.

Reply:
Ok